lib/Controller/SocialApiController.php

```diff
@@ -0,0 +1,208 @@
+
+use OCP\AppFramework\ApiController;
+use OCP\AppFramework\Http;
+use OCP\AppFramework\Http\DataResponse;
+use OCP\AppFramework\Http\JSONResponse;
+use OCP\AppFramework\Http\TemplateResponse;
+// use OCP\IInitialStateService;
+use OCP\IConfig;
+use OCP\Contacts\IManager;
+use OCP\L10N\IFactory;
+use OCP\IRequest;
+
+
+class SocialApiController extends ApiController {
+
+	protected $appName;
+
+	// /** @var IInitialStateService */
+	// private $initialStateService;
+
+	/** @var IFactory */
+	private $languageFactory;
+	/** @var IManager */
+	private  $manager;
+	/** @var IConfig */
+	private  $config;
+
+	public function __construct(string $AppName,
+								IRequest $request,
+								IManager $manager,
+								IConfig $config,
+								// IInitialStateService $initialStateService,
+								IFactory $languageFactory) {
+		parent::__construct($AppName, $request);
+
+		$this->appName = $AppName;
+		// $this->initialStateService = $initialStateService;
+		$this->languageFactory = $languageFactory;
+		$this->manager = $manager;
+		$this->config = $config;
+	}
+
+
+	/**
+	 * @NoAdminRequired
+	 *
+	 * generate download url for a social entry (based on type of data requested)
+	 */
+	protected function getSocialConnector(array $socialentry, string $type) : ?string {
+		foreach ($socialentry as $network => $candidate) {
+			$connector = null;
+			$valid = false;
+	
+			// get profile-id
+			switch (strtolower($network)) {
+				case "facebook":
+					$candidate = basename($candidate);
+					if (!ctype_digit($candidate)) {
+						// TODO: determine facebook profile id from username
+						break;
+					}
+					$valid = true;
+					break;
+			}
+			if ($valid) {
+				// build connector
+				switch (strtolower($network)) {
+					case "facebook":
+						switch ($type) {
+							case "avatar":
+								$connector = "https://graph.facebook.com/" . ($candidate) . "/picture?width=720";
+								break;
+							default:
+								break;
+						}
+						break;
+				}
+				
+				// return first valid connector:
+				if ($connector) { return ($connector); }
+			}
+		}
+		
+		// reached only if no valid connectors found
+		return null;
+	}
+
+	/**
+	 * @NoAdminRequired
+	 *
+	 * Retrieves social profile data for a contact
+	 */
+	public function fetch(string $addressbookId, string $contactId, string $type) : JSONResponse {
+
+		$url = null;
+		$response = new JSONResponse(array());
+		$response->setStatus(404);
+
+		try {
+			// get corresponding addressbook
+			$addressBooks = $this->manager->getUserAddressBooks();
+			$addressBook = null;
+			foreach($addressBooks as $ab) {
+				if ($ab->getUri() === $addressbookId) {
+					$addressBook = $ab;
+				}
+			}
+			if (is_null($addressBook)) {
+				$response->setStatus(500);
+				return $response; 
+			}
+			
+			// search contact in that addressbook
+			$contact = $addressBook->search($contactId, ['UID'], [])[0];
+			if (is_null($addressBook)) {
+				$response->setStatus(500);
+				return $response; 
```

Comment:
```suggestion
				return new Http\JSONResponse([], Http::STATUS_NOT_FOUND);
```

Cleaner to directly return the appropriate response, no need to build it and edit it along the script :)